mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/RabbitMQEventBus.java

```diff
@@ -36,34 +36,54 @@
 import reactor.rabbitmq.Sender;
 import reactor.rabbitmq.SenderOptions;
 
-public class RabbitMQEventBus implements EventBus {
+class RabbitMQEventBus implements EventBus {
     static final String MAILBOX_EVENT = "mailboxEvent";
     static final String MAILBOX_EVENT_EXCHANGE_NAME = MAILBOX_EVENT + "-exchange";
 
-    private final Sender sender;
-    private final GroupRegistrationHandler groupRegistrationHandler;
-    private final EventDispatcher eventDispatcher;
+    private final Mono<Connection> connectionMono;
+    private final EventSerializer eventSerializer;
+    private final AtomicBoolean isRunning;
 
-    RabbitMQEventBus(RabbitMQConnectionFactory rabbitMQConnectionFactory, EventSerializer eventSerializer) {
-        Mono<Connection> connectionMono = Mono.fromSupplier(rabbitMQConnectionFactory::create).cache();
+    private GroupRegistrationHandler groupRegistrationHandler;
+    private final KeyRegistrationHandler keyRegistrationHandler;
+    private EventDispatcher eventDispatcher;
+    private Sender sender;
+
+    RabbitMQEventBus(RabbitMQConnectionFactory rabbitMQConnectionFactory, EventSerializer eventSerializer, RoutingKeyConverter routingKeyConverter) {
+        this.connectionMono = Mono.fromSupplier(rabbitMQConnectionFactory::create).cache();
         this.sender = RabbitFlux.createSender(new SenderOptions().connectionMono(connectionMono));
         this.groupRegistrationHandler = new GroupRegistrationHandler(eventSerializer, sender, connectionMono);
+        this.keyRegistrationHandler = new KeyRegistrationHandler(eventSerializer, sender, connectionMono, routingKeyConverter);
         this.eventDispatcher = new EventDispatcher(eventSerializer, sender);
+        this.eventSerializer = eventSerializer;
+        isRunning = new AtomicBoolean(false);
     }
 
     public void start() {
-        eventDispatcher.start();
+        if (!isRunning.get()) {
+            sender = RabbitFlux.createSender(new SenderOptions().connectionMono(connectionMono));
```

Comment:
Why do we re-create the sender here, and also in the constructor?

If so, maybe you don't want to create it in the constructor... This will also force you to re-create the KeyRegistrationHandler that you forgot...